core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -106,11 +106,19 @@ object EntityDecoder extends EntityDecoderInstances {
       msg.headers.get(`Content-Type`) match {
         case Some(contentType) =>
           if (a.matchesMediaType(contentType.mediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMismatch(actual, expected) =>
+              MediaTypeMismatch(actual, expected ++ a.consumes)
+            case other => other
+          }
 
         case None =>
           if (a.matchesMediaType(UndefinedMediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMissing(expected) =>
+              MediaTypeMissing(expected ++ a.consumes)
```

Comment:
I meant our code here has the expectation of only receiving one of the two. I agree that externally, no such guarantee is made, but in the implementation of `OrDec`, we do not handle `MediaTypeMismatch` in the `None` case because we "know" it will not happen and vice-versa. But if it were to happen (since we have no guarantee it won't), then we would have a subtle bug where in certain cases the list of expected media-types is not complete. I guess we could handle both error types in both calls to `leftMap` but that does not seem ideal either.